chalice/deploy/planner.py

```diff
@@ -0,0 +1,178 @@
+import json
+
+import attr
+from typing import List, Dict, Any, Optional, Union, Tuple, cast  # noqa
+
+from chalice.utils import OSUtils  # noqa
+from chalice.deploy import models
+from chalice.awsclient import TypedAWSClient, ResourceDoesNotExistError  # noqa
+
+
+class RemoteState(object):
+    def __init__(self, client):
+        # type: (TypedAWSClient) -> None
+        self._client = client
+        self._cache = {}  # type: Dict[Tuple[str, str], bool]
+
+    def _cache_key(self, resource):
+        # type: (models.ManagedModel) -> Tuple[str, str]
+        return (resource.resource_type, resource.resource_name)
+
+    def resource_exists(self, resource):
+        # type: (models.ManagedModel) -> bool
+        key = self._cache_key(resource)
+        if key in self._cache:
+            return self._cache[key]
+        if isinstance(resource, models.ManagedIAMRole):
+            result = self._resource_exists_iam_role(resource)
+        elif isinstance(resource, models.LambdaFunction):
+            result = self._resource_exists_lambda_function(resource)
+        self._cache[key] = result
+        return result
+
+    def _resource_exists_lambda_function(self, resource):
+        # type: (models.LambdaFunction) -> bool
+        return self._client.lambda_function_exists(resource.function_name)
+
+    def _resource_exists_iam_role(self, resource):
+        # type: (models.ManagedIAMRole) -> bool
+        try:
+            self._client.get_role_arn_for_name(resource.role_name)
+            return True
+        except ResourceDoesNotExistError:
+            return False
+
+    def get_remote_model(self, resource):
```

Comment:
Yeah I do think we should do an optimization pass and add caching.  It might be worth doing this as a separate PR, but I can at least add a TODO so we don't forget.

Reply:
Yeah a TODO won't hurt as a reminder.